backend_new/src/services/auth.service.ts

```diff
@@ -0,0 +1,335 @@
+        mfaCodeUpdatedAt: new Date(),
+        phoneNumber: user.phoneNumber,
+      },
+      where: {
+        id: user.id,
+      },
+    });
+
+    if (dto.mfaType === MfaType.email) {
+      // TODO: email service (https://github.com/bloom-housing/bloom/pull/3607)
+    } else if (dto.mfaType === MfaType.sms) {
+      await this.smsService.sendMfaCode(user.phoneNumber, mfaCode);
+    }
+
+    return dto.mfaType === MfaType.email
+      ? { email: user.email, phoneNumberVerified: user.phoneNumberVerified }
+      : {
+          phoneNumber: user.phoneNumber,
+          phoneNumberVerified: user.phoneNumberVerified,
+        };
+  }
+
+  /*
+    updates a user's password and logs them in
+  */
+  async updatePassword(
+    dto: UpdatePassword,
+    res: Response,
+  ): Promise<SuccessDTO> {
+    const user = await this.prisma.userAccounts.findFirst({
+      where: { resetToken: dto.token },
+    });
+
+    if (!user) {
+      throw new NotFoundException(
+        `user resetToken: ${dto.token} was requested but not found`,
+      );
+    }
+
+    const token: IdDTO = verify(dto.token, process.env.APP_SECRET) as IdDTO;
+
+    if (token.id !== user.id) {
+      throw new UnauthorizedException(
+        `resetToken ${dto.token} does not match user ${user.id}'s reset token (${user.resetToken})`,
+      );
+    }
+
+    await this.prisma.userAccounts.update({
+      data: {
+        passwordHash: await passwordToHash(dto.password),
+        passwordUpdatedAt: new Date(),
+        resetToken: null,
+      },
+      where: {
+        id: user.id,
+      },
+    });
+
+    return await this.setCredentials(res, mapTo(User, user));
+  }
+
+  /*
+    confirms a user and logs them in
+  */
+  async confirmUser(dto: Confirm, res?: Response): Promise<SuccessDTO> {
+    const token = verify(dto.token, process.env.APP_SECRET) as IdAndEmail;
+
+    let user = await this.userService.findUserOrError(
+      { userId: token.id },
+      false,
+    );
+
+    if (user.confirmationToken !== dto.token) {
+      throw new BadRequestException(
+        `Confirmation token mismatch for user stored: ${user.confirmationToken}, incoming token: ${dto.token}`,
+      );
+    }
+
+    const data: Prisma.UserAccountsUpdateInput = {
+      confirmedAt: new Date(),
+      confirmationToken: null,
+    };
+
+    if (dto.password) {
+      data.passwordHash = await passwordToHash(dto.password);
+      data.passwordUpdatedAt = new Date();
+    }
+
+    if (token.email) {
+      data.email = token.email;
+    }
+
+    user = await this.prisma.userAccounts.update({
+      data,
+      where: {
+        id: user.id,
+      },
+    });
+
+    return await this.setCredentials(res, mapTo(User, user));
+  }
+
+  /*
+    generates a numeric mfa code
+  */
+  generateMfaCode() {
+    let out = '';
+    const characters = '0123456789';
+    for (let i = 0; i < Number(process.env.MFA_CODE_LENGTH); i++) {
+      out += characters.charAt(Math.floor(Math.random() * characters.length));
```

Comment:
I've been using the crypto `randomInt` function instead of the `Math.floor(Math.random()...`.

Reply:
I much prefer that! switched!